torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,110 @@
+from typing import Any, Callable, List, Optional
+
+import torch
+import torch.distributed as dist
+from torch.distributed.optim import ZeroRedundancyOptimizer
+from torch.distributed.optim.zero_redundancy_optimizer import _OverlapStatus
+from torch.nn.parallel.distributed import DistributedDataParallel
+
+
+def hook_then_zero_step(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future],
+    ddp: DistributedDataParallel,
+    zero: ZeroRedundancyOptimizer,
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    r"""
+    Modifies the given ``hook`` to additionally perform a partial
+    :class:`ZeroRedundancyOptimizer` :meth:`step` using the gradients in the
+    :class:`DistributedDataParallel` gradient bucket provided by the ``hook``.
+
+    Arguments:
+        hook (Any * dist.GradBucket -> torch.futures.Future): the hook to
+            modify.
+        ddp (DistributedDataParallel): the :class:`DistributedDataParallel`
+            instance to use.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to use.
+
+    Returns:
+        The modified hook.
+    """
+    assert zero._is_functional_optim, "ZeroRedundancyOptimizer must be " \
+        "constructed with a functional optimizer class to use this function"
+
+    def hook_then_zero_fn(
+        state,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step` using
+        the gradients in that bucket.
+
+        Arguments:
+            state: any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+
+        def zero_step(_) -> List[torch.Tensor]:
+            r"""
+            Performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step`
+            using the gradients in the given :class:`DistributedDataParallel`
+            gradient bucket.
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
```

Comment:
This means we don't do optimizer step at all before buckets are rebuilt?

Reply:
Yes. Sorry that I have not had a chance to write the PR text yet explaining the changes. This overlap implementation requires two dummy iterations to warm up, and the training loop still requires `ZeroRedundancyOptimizer.step()` to be called. It becomes a no-op once warmed up.

Reply:
(can be done in followup PRs) is it possible to let ZeRO to adapt to DDP's rebuilt buckets?

Reply:
Here, we might also need to get the Future as argument, and return `[Future.wait()]`, otherwise `wait()` on the child future might not wait for the allreduce to finish.

Reply:
Commenting for completeness:

The `Future.wait()` here is not needed if using a single stream for computation since `then()` performs the needed synchronization to guarantee the all-reduce finishes. However, it is needed if the optimizer computation occurs in a separate stream. (Note: The API has changed so that the return should look like `return bucket.get_tensor()` now. However, the current all-reduce hook returns a singleton `List[Tensor]` still, so the call would look like `return fut.wait()[0]`, where `fut` is the argument to `zero_step()`.)